HashMap uses structural sharing to achieve sub-linear concatenation

joshlemer opened the pull request:
Also depends on #7118 in order to use mutability of the MapNodes.  
If this goes through, I will gladly do the same for HashSets.

joshlemer commented:
Relevant method is here https://github.com/scala/scala/pull/7192/files#diff-43417c3d2b66ce19a2d4981ab39c0ae5R108

joshlemer commented:
By my calculations, this constitutes a **3x - 5x improvement**!  
  
```
[info] Benchmark                   (size)  (stringsOnly)  (useMissingValues)  Mode  Cnt         Score         Error  Units
[info] HashMapBenchmark.concatNew      10          false                true  avgt   20       397.815 ±      14.853  ns/op
[info] HashMapBenchmark.concatNew    1000          false                true  avgt   20     88685.365 ±    1437.678  ns/op
[info] HashMapBenchmark.concatNew  100000          false                true  avgt   20  13637356.687 ±  155080.980  ns/op
[info] HashMapBenchmark.concatOld      10          false                true  avgt   20      1944.114 ±      31.764  ns/op
[info] HashMapBenchmark.concatOld    1000          false                true  avgt   20    303826.136 ±   12042.900  ns/op
[info] HashMapBenchmark.concatOld  100000          false                true  avgt   20  39177164.361 ± 2559757.974  ns/op
```

mkeskells commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -570,43 +860,69 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
+    val index = indexOf(key)
+    if (index >= 0) {
+
+      if (content(index)._2.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef]) {
```
we should check if the key is eq here as well, in line with the general guidance on map key handling - looks like the old implementation didn't do this though

mkeskells commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -570,43 +860,69 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
+private final class HashCollisionMapNode[K, +V ](
+  val originalHash: Int,
+  val hash: Int,
+  var content: Vector[(K, V @uV)]
```
wouldn't it be better to maintain the content as Array/Vector[AnyRef]
this reduces the allocation of tuples, and is more similar to BitmapIndexedMapNode

I am not convinced that Vector is providing and structural sharing here (maybe where we have a collision that has >32 collisions already and we add another one)
If structural sharing is not useful then shouln't this be an Array to reduce the footprint. Maybe its a sperate PR though as the change is orthogonal to this one

joshlemer replied:
I think this is like you say really important, and worthy of an other discussion

joshlemer replied:
@mkeskells see https://github.com/scala/bug/issues/11164

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+                      val rightOriginalHash = rightBm.getHash(rightDataIdx)
+                      new BitmapIndexedMapNode(0, 0, Array(), Array(), 0)
+                        .updated(leftBm.getKey(leftDataIdx), leftBm.getValue(leftDataIdx), leftOriginalHash, improve(leftOriginalHash), nextShift)
+                        .updated(rightBm.getKey(rightDataIdx), rightBm.getValue(rightDataIdx), rightOriginalHash, improve(rightOriginalHash), nextShift)
```
Looks like this could be optimized to create and prefill the content array of the correct size, rather than use `updated` to do the same.

joshlemer replied:
Yes, I took the lazy way out haha

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+
+    that match {
+      case hm: HashMap[K, V1] =>
+        var newHash = cachedJavaKeySetHashCode + hm.cachedJavaKeySetHashCode
```
Add a comment here to explain how this ends up at the right value. e.g. 

"We start with the assumption that the maps are distinct, and then subtract the hash code of keys in  `this` are overwritten by keys in `that`."

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
   }
 
   override def concat[V1 >: V](that: scala.IterableOnce[(K, V1)]): HashMap[K, V1] = {
```
I'm assuming that when we [optimize merge](https://github.com/scala/bug/issues/11077) we'll be able to share the implementation with `concat`. Do you think it's worth doing `merge` in this PR so we can shake out a good structure?

joshlemer replied:
I actually don't think we'll be able to share implementation with `concat` -- the result of the `merge` function is not constrained to `leftKey | rightKey`, it can be any `K`, so can fall anywhere in the resulting map's trie structure.

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+        var newHash = cachedJavaKeySetHashCode + hm.cachedJavaKeySetHashCode
+
+        /** Recursively, immutably concatenates two MapNodes, node-by-node as to visit the least number of nodes possible */
+        def concat(left: MapNode[K, V1], right: MapNode[K, V1], shift: Int): MapNode[K, V1] =
```
We might want to make this constant stack to be consistent with the other trie traversals.

Do you think we could extract some analog of `ChampBaseIterator` to capture the parts of zippy traversal that are independant of the `Map`/`Set` distinction, and implement this in terms of that?

joshlemer replied:
Yeah I've been thinking a lot about that. That is probably the right way to go so that these classes don't have a dozen slightly different intricate bit fiddling implementations, I just have a hard time thinking how it would look

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+                      val rightImproved = improve(rightOriginalHash)
+
+                      if (n.containsKey(rightKey, rightOriginalHash, rightImproved, nextShift)) {
+                        newHash -= improve(rightBm.getHash(rightDataIdx))
```
Could we base this hash adjustment on `newNode.size == n.size` instead?

joshlemer replied:
Great idea!

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+                      n.updated(rightKey, rightValue, rightOriginalHash, rightImproved, nextShift)
+                    }
+
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
```
If `rightKey`/`rigthValue` are already in (by reference equality) the left node, we'll find that `newNode eq n` here. Could that be the basis for further optimizations, where we could structurally share in paths where the left and right were both present, but pointed at identical payloads and at identical sub-nodes (the result of prior structural sharing).

```scala
val a = "a"; val b = "b", ...
val m1 = Map(a -> a, b -> b, ...)
val m2 = Map(a -> a, b -> b, ...)
val m3 = m1.concat(m2)
assert(m1 eq m2)
```

joshlemer replied:
If I understand you correctly, are you suggesting that each node-wise concat first does an entire sweep of its data and nodes, and if it finds that for all of them `leftNodeOrData eq rightNodeOrData`, then just return, say `leftNode`?

joshlemer committed:
HashMap#concat uses structural sharing to achieve sublinear performance

Apply @retronym's suggestions

Revert breaking HashMap#concat changes

In HashMap#concat, rely on before/after size when merging leftNode+rightData to determine if newHash should be changed

Lots of progress on more Map concat micro optimizations

Remove tzeroes and lzeroes

Remove unused if/else clause

Add comment explaining that hash collisions are never at the same level as bitmapIndexedMapNodes

use mergeTwoKeyValPairs to create new node in HashMap#concat

Remove testing code

joshlemer commented:
Here are some updated, gc profiled benchmarks
```
[info] # Run complete. Total time: 00:06:11
[info] 
[info] Benchmark                                                    (size)  (stringsOnly)  (useMissingValues)  Mode  Cnt       Score       Error   Units
[info] HashMapBenchmark.concatNew                                       10          false                true  avgt   20     348.407 ±     9.420   ns/op
[info] HashMapBenchmark.concatNew:·gc.alloc.rate                        10          false                true  avgt   20     598.518 ±    15.156  MB/sec
[info] HashMapBenchmark.concatNew:·gc.alloc.rate.norm                   10          false                true  avgt   20     328.001 ±     0.001    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space               10          false                true  avgt   20     603.696 ±    25.401  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space.norm          10          false                true  avgt   20     330.766 ±     9.553    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space           10          false                true  avgt   20       0.095 ±     0.030  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space.norm      10          false                true  avgt   20       0.052 ±     0.017    B/op
[info] HashMapBenchmark.concatNew:·gc.count                             10          false                true  avgt   20     307.000              counts
[info] HashMapBenchmark.concatNew:·gc.time                              10          false                true  avgt   20     172.000                  ms
[info] HashMapBenchmark.concatNew                                      100          false                true  avgt   20    8800.484 ±   120.176   ns/op
[info] HashMapBenchmark.concatNew:·gc.alloc.rate                       100          false                true  avgt   20     433.011 ±     9.378  MB/sec
[info] HashMapBenchmark.concatNew:·gc.alloc.rate.norm                  100          false                true  avgt   20    6000.014 ±    85.528    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space              100          false                true  avgt   20     433.219 ±    14.766  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space.norm         100          false                true  avgt   20    6002.916 ±   179.719    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space          100          false                true  avgt   20       0.087 ±     0.033  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space.norm     100          false                true  avgt   20       1.210 ±     0.468    B/op
[info] HashMapBenchmark.concatNew:·gc.count                            100          false                true  avgt   20     356.000              counts
[info] HashMapBenchmark.concatNew:·gc.time                             100          false                true  avgt   20     185.000                  ms
[info] HashMapBenchmark.concatNew                                     1000          false                true  avgt   20   77209.319 ±  1748.751   ns/op
[info] HashMapBenchmark.concatNew:·gc.alloc.rate                      1000          false                true  avgt   20     640.018 ±    14.387  MB/sec
[info] HashMapBenchmark.concatNew:·gc.alloc.rate.norm                 1000          false                true  avgt   20   77800.127 ±     0.250    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space             1000          false                true  avgt   20     639.595 ±    20.876  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Eden_Space.norm        1000          false                true  avgt   20   77755.414 ±  2019.348    B/op
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space         1000          false                true  avgt   20       0.038 ±     0.018  MB/sec
[info] HashMapBenchmark.concatNew:·gc.churn.PS_Survivor_Space.norm    1000          false                true  avgt   20       4.676 ±     2.202    B/op
[info] HashMapBenchmark.concatNew:·gc.count                           1000          false                true  avgt   20     348.000              counts
[info] HashMapBenchmark.concatNew:·gc.time                            1000          false                true  avgt   20     181.000                  ms
[info] HashMapBenchmark.concatOld                                       10          false                true  avgt   20    1811.681 ±    98.218   ns/op
[info] HashMapBenchmark.concatOld:·gc.alloc.rate                        10          false                true  avgt   20    1602.095 ±    72.758  MB/sec
[info] HashMapBenchmark.concatOld:·gc.alloc.rate.norm                   10          false                true  avgt   20    4556.003 ±    53.455    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space               10          false                true  avgt   20    1603.234 ±    88.234  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space.norm          10          false                true  avgt   20    4558.464 ±   131.061    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space           10          false                true  avgt   20       0.104 ±     0.026  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space.norm      10          false                true  avgt   20       0.295 ±     0.072    B/op
[info] HashMapBenchmark.concatOld:·gc.count                             10          false                true  avgt   20     296.000              counts
[info] HashMapBenchmark.concatOld:·gc.time                              10          false                true  avgt   20     172.000                  ms
[info] HashMapBenchmark.concatOld                                      100          false                true  avgt   20   15076.857 ±   320.297   ns/op
[info] HashMapBenchmark.concatOld:·gc.alloc.rate                       100          false                true  avgt   20    1256.976 ±    17.557  MB/sec
[info] HashMapBenchmark.concatOld:·gc.alloc.rate.norm                  100          false                true  avgt   20   29852.025 ±   481.094    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space              100          false                true  avgt   20    1253.790 ±    40.483  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space.norm         100          false                true  avgt   20   29773.066 ±   910.525    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space          100          false                true  avgt   20       0.066 ±     0.029  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space.norm     100          false                true  avgt   20       1.576 ±     0.694    B/op
[info] HashMapBenchmark.concatOld:·gc.count                            100          false                true  avgt   20     321.000              counts
[info] HashMapBenchmark.concatOld:·gc.time                             100          false                true  avgt   20     181.000                  ms
[info] HashMapBenchmark.concatOld                                     1000          false                true  avgt   20  241860.093 ±  3299.076   ns/op
[info] HashMapBenchmark.concatOld:·gc.alloc.rate                      1000          false                true  avgt   20    1210.613 ±    16.035  MB/sec
[info] HashMapBenchmark.concatOld:·gc.alloc.rate.norm                 1000          false                true  avgt   20  461640.395 ±     0.773    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space             1000          false                true  avgt   20    1215.668 ±    37.330  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Eden_Space.norm        1000          false                true  avgt   20  463554.593 ± 12458.997    B/op
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space         1000          false                true  avgt   20       0.056 ±     0.029  MB/sec
[info] HashMapBenchmark.concatOld:·gc.churn.PS_Survivor_Space.norm    1000          false                true  avgt   20      21.371 ±    11.339    B/op
[info] HashMapBenchmark.concatOld:·gc.count                           1000          false                true  avgt   20     362.000              counts
[info] HashMapBenchmark.concatOld:·gc.time                            1000          false                true  avgt   20     185.000                  ms
```

joshlemer commented:
Over the weekend I worked on the following optimizations:

* When possible, when we know that `leftNode` will not contribute any data to the result, we just return `rightNode`. This can only be done at the top level though, because in lower levels, we need to at least visit each immediate subnode to inspect which elements overlap in order to compute the new hash.
* In bitmapIndexedMapNode merging we only visit from `minimumIndex` to `maximumIndex`, rather than from `0` to `31`, by computing the minimum /maximum index from the trailing/leading zeros in the combined left/right data/node maps.
* bit position is not computed from index each time, it is instead started at `minimumBitPos`, and on each iteration, it is bit shifted by one `bitpos = bitpos << 1`, in fact there no longer is a non-bitpos index in the iterations. 
* At each level, at the end of the recursive `concat` we determine if any changes have been made, such that the result is not equal to `right` (by flagging such changes in the boolean `anyChangesMadeSoFar`. If no changes have been made, then return `right`. This will allow the newly constructed node to be freed.

retronym commented:
/synch

retronym reviewed (approved):
Let's book the progress here, copy this across to HashSet, and address the remaining ideas in future PRs.
